### PR TITLE
qt: Add UI to manage emulated user profiles

### PR DESCRIPTION
**Pull request by DarkLordZach**
tl;dr multiple users and user icons

This adds a UI to manage emulated users along with config support for multiple users and loading of custom icons for users.

For icons, yuzu looks in `%YUZU_DIR%/config/users/<username>.jpg`. It must be a JPEG less than 20kb in size due to switch restrictions. Should no suitable icon be found, a gray background will be used.

The UI allows adding, renaming, and deleting users. When a new user is made, the UUID will be randomly generated. 

Additionally, the right-click open save data menu will let you pick the user.

settings
(NSO NES - icon and name in top left corner)
ppt
(Puyo Puyo - custom name in blue box in bottom left corner)

### CODE REVIEW

File: src/yuzu/configuration/configure_system.cpp [hunk 1]
```diff
@@ -92,3 +189,67 @@ void ConfigureSystem::refreshConsoleID() {
...
 }
+
+void ConfigureSystem::SelectUser(const QModelIndex& index) {
+    Settings::values.current_user =
+        std::clamp<std::size_t>(index.row(), 0, Settings::values.users.size() - 1);
+
+    UpdateCurrentUser();
+
+    if (Settings::values.users.size() >= 2)
+        ui->pm_remove->setEnabled(true);
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: This can collapse to:
```cpp
ui->pm_remove->setEnabled(Settings::values.users.size() >= 2);
```

File: src/yuzu/configuration/configure_system.cpp [hunk 2]
```diff
@@ -92,3 +189,67 @@ void ConfigureSystem::refreshConsoleID() {
...
+        ui->pm_remove->setEnabled(false);
+
+    ui->pm_rename->setEnabled(true);
+}
+
+void ConfigureSystem::AddUser() {
+    Service::Account::UUID uuid;
+    uuid.Generate();
+
+    bool ok = false;
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: Shouldn't the value of `ok` be checked after the `getText()` call below.

File: src/yuzu/configuration/config.cpp [hunk 3]
```diff
@@ -122,7 +122,25 @@ void Config::ReadValues() {
 
     qt_config->beginGroup("System");
     Settings::values.use_docked_mode = qt_config->value("use_docked_mode", false).toBool();
-    Settings::values.username = qt_config->value("username", "yuzu").toString().toStdString();
+
+    Settings::values.users.clear();
```
Comment (lioncash) [inline src/yuzu/configuration/config.cpp]: The code to load all the users should be placed within its own function.

File: src/yuzu/configuration/config.cpp [hunk 4]
```diff
@@ -247,7 +265,19 @@ void Config::SaveValues() {
 
     qt_config->beginGroup("System");
     qt_config->setValue("use_docked_mode", Settings::values.use_docked_mode);
-    qt_config->setValue("username", QString::fromStdString(Settings::values.username));
+    qt_config->setValue("current_user", Settings::values.current_user);
```
Comment (lioncash) [inline src/yuzu/configuration/config.cpp]: Ditto here, but with the saving code.

File: src/yuzu/configuration/config.cpp [hunk 5]
```diff
@@ -247,7 +265,19 @@ void Config::SaveValues() {
...
+
+    qt_config->beginWriteArray("users", Settings::values.users.size());
+    for (std::size_t i = 0; i < Settings::values.users.size(); ++i) {
+        qt_config->setArrayIndex(i);
```
Comment (lioncash) [inline src/yuzu/configuration/config.cpp]: `i` should be casted to an `int`, otherwise this will cause a truncation warning.

File: src/yuzu/configuration/configure_system.cpp [hunk 6]
```diff
@@ -2,7 +2,12 @@
 // Licensed under GPLv2 or any later version
 // Refer to the license.txt file included.
 
+#include <QGraphicsItem>
+#include <QList>
 #include <QMessageBox>
+#include <qinputdialog.h>
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: `<QInputDialog>`

File: src/yuzu/configuration/configure_system.cpp [hunk 7]
```diff
@@ -32,23 +48,104 @@ ConfigureSystem::ConfigureSystem(QWidget* parent) : QWidget(parent), ui(new Ui::
...
+                       QPixmap icon{icon_url};
+
+                       if (!icon) {
+                           icon.fill(QColor::fromRgb(0, 0, 0));
+                           icon.loadFromData(backup_jpeg.data(), backup_jpeg.size());
+                       }
+
+                       return QList{new QStandardItem{
+                           icon.scaled(64, 64, Qt::IgnoreAspectRatio, Qt::SmoothTransformation),
+                           QString::fromStdString(user.first + "\n" + user.second.Format())}};
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: '\n'

File: src/yuzu/configuration/configure_system.cpp [hunk 8]
```diff
@@ -32,23 +48,104 @@ ConfigureSystem::ConfigureSystem(QWidget* parent) : QWidget(parent), ui(new Ui::
...
+
     ui->combo_language->setCurrentIndex(Settings::values.language_index);
+
+    item_model->removeRows(0, item_model->rowCount());
+    list_items.clear();
+
+    std::transform(Settings::values.users.begin(), Settings::values.users.end(),
+                   std::back_inserter(list_items),
+                   [](const std::pair<std::string, Service::Account::UUID>& user) {
+                       const auto icon_url = QString::fromStdString(
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: This lambda is essentially partially duplicating the `UpdateCurrentUser()` function.

File: src/yuzu/configuration/configure_system.cpp [hunk 9]
```diff
@@ -92,3 +189,67 @@ void ConfigureSystem::refreshConsoleID() {
...
+    uuid.Generate();
+
+    bool ok = false;
+    const auto username =
+        QInputDialog::getText(this, tr("Enter Username"), tr("Enter a username for the new user:"),
+                              QLineEdit::Normal, QString(), &ok);
+
+    Settings::values.users.emplace_back(username.toStdString(), uuid);
+
+    setConfiguration();
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: It seems really janky to need to completely reload the entire user list if only one new user is added.

File: src/yuzu/configuration/configure_system.cpp [hunk 10]
```diff
@@ -92,3 +189,67 @@ void ConfigureSystem::refreshConsoleID() {
...
+    const auto new_username = QInputDialog::getText(
+        this, tr("Enter Username"), tr("Enter a new username:"), QLineEdit::Normal,
+        QString::fromStdString(Settings::values.users[user].first), &ok);
+
+    if (!ok)
+        return;
+
+    Settings::values.users[user].first = new_username.toStdString();
+
+    setConfiguration();
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: Especially so for renaming.

File: src/yuzu/configuration/configure_system.cpp [hunk 11]
```diff
@@ -92,3 +189,67 @@ void ConfigureSystem::refreshConsoleID() {
...
+
+    if (confirm == QMessageBox::No)
+        return;
+
+    if (Settings::values.current_user == tree_view->currentIndex().row())
+        Settings::values.current_user = 0;
+
+    Settings::values.users.erase(user);
+
+    setConfiguration();
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: Ditto for removing a user.

File: src/yuzu/configuration/configure_system.h [hunk 12]
```diff
@@ -5,6 +5,11 @@
 #pragma once
 
 #include <memory>
+#include <QGraphicsScene>
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.h]: Many of these can be forward declarations.

File: src/yuzu/main.cpp [hunk 13]
```diff
@@ -725,12 +725,33 @@ void GMainWindow::OnGameListOpenFolder(u64 program_id, GameListOpenTarget target
         open_target = "Save Data";
         const std::string nand_dir = FileUtil::GetUserPath(FileUtil::UserPath::NANDDir);
         ASSERT(program_id != 0);
-        // TODO(tech4me): Update this to work with arbitrary user profile
-        // Refer to core/hle/service/acc/profile_manager.cpp ProfileManager constructor
-        constexpr u128 user_id = {1, 0};
+
+        QStringList list{};
```
Comment (lioncash) [inline src/yuzu/main.cpp]: `{}` isn't necessary here.

File: src/yuzu/main.cpp [hunk 14]
```diff
@@ -725,12 +725,33 @@ void GMainWindow::OnGameListOpenFolder(u64 program_id, GameListOpenTarget target
         open_target = "Save Data";
         const std::string nand_dir = FileUtil::GetUserPath(FileUtil::UserPath::NANDDir);
         ASSERT(program_id != 0);
-        // TODO(tech4me): Update this to work with arbitrary user profile
-        // Refer to core/hle/service/acc/profile_manager.cpp ProfileManager constructor
-        constexpr u128 user_id = {1, 0};
+
```
Comment (lioncash) [inline src/yuzu/main.cpp]: This code should probably go within its own function.

File: src/yuzu/configuration/configure_system.cpp [hunk 15]
```diff
@@ -32,23 +48,104 @@ ConfigureSystem::ConfigureSystem(QWidget* parent) : QWidget(parent), ui(new Ui::
...
+
+void ConfigureSystem::UpdateCurrentUser() {
+    const auto& current_user = Settings::values.users[Settings::values.current_user];
+    const auto icon_url =
+        QString::fromStdString(FileUtil::GetUserPath(FileUtil::UserPath::ConfigDir) + "users" +
+                               DIR_SEP + current_user.first + ".jpg");
+    QPixmap icon{icon_url};
+
+    if (!icon) {
+        icon.fill(QColor::fromRgb(0, 0, 0));
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: just calling `icon.fill(Qt::black);` will perform the same behavior.

File: src/core/hle/service/am/am.cpp [hunk 16]
```diff
@@ -670,16 +672,16 @@ IApplicationFunctions::IApplicationFunctions() : ServiceFramework("IApplicationF
...
-        0,    0,    0,    0,    // User Id (word 1)
-        0,    0,    0,    0,    // User Id (word 2)
-        0,    0,    0,    0     // User Id (word 3)
-    }};
+    };
+
+    std::vector<u8> buffer(POP_LAUNCH_PARAMETER_BUFFER_SIZE);
 
-    std::vector<u8> buffer(data.begin(), data.end());
+    std::memcpy(buffer.data(), header_data.data(), header_data.size());
```
Comment (lioncash) [inline src/core/hle/service/am/am.cpp]: `<cstring>` needs to be included.

File: src/core/hle/service/acc/acc.cpp [hunk 17]
```diff
@@ -87,22 +106,38 @@ class IProfile final : public ServiceFramework<IProfile> {
...
+
+        if (image == nullptr)
+            rb.Push<u32>(backup_jpeg_size);
+        else
+            rb.Push<u32>(std::min<u32>(image->GetSize(), MAX_JPEG_IMAGE_SIZE));
     }
 
     const ProfileManager& profile_manager;
     UUID user_id; ///< The user id this profile refers to.
+    std::unique_ptr<FileUtil::IOFile> image = nullptr;
```
Comment (lioncash) [inline src/core/hle/service/acc/acc.cpp]: Why does this need to contain a file handle the whole time it's active rather than just the data within the file?. At least that way avoids holding a handle open to a file.

File: src/core/hle/service/acc/acc.cpp [hunk 18]
```diff
@@ -87,22 +106,38 @@ class IProfile final : public ServiceFramework<IProfile> {
...
+
         IPC::ResponseBuilder rb{ctx, 3};
         rb.Push(RESULT_SUCCESS);
-        rb.Push<u32>(jpeg_size);
+
+        if (image == nullptr) {
+            ctx.WriteBuffer(backup_jpeg);
+            rb.Push<u32>(backup_jpeg_size);
+        } else {
+            const auto size = std::min<u32>(image->GetSize(), MAX_JPEG_IMAGE_SIZE);
```
Comment (lioncash) [inline src/core/hle/service/acc/acc.cpp]: `<algorithm>` needs to be included.

File: src/core/settings.h [hunk 19]
```diff
@@ -113,7 +114,8 @@ static const std::array<const char*, NumAnalogs> mapping = {{
 struct Values {
     // System
     bool use_docked_mode;
-    std::string username;
+    int current_user;
+    std::vector<std::pair<std::string, Service::Account::UUID>> users;
```
Comment (lioncash) [inline src/core/settings.h]: Fairly sure it's not a good idea to dump service internals into the main settings header.

File: src/yuzu/configuration/configure_system.cpp [hunk 20]
```diff
@@ -32,23 +48,104 @@ ConfigureSystem::ConfigureSystem(QWidget* parent) : QWidget(parent), ui(new Ui::
...
 void ConfigureSystem::setConfiguration() {
     enabled = !Core::System::GetInstance().IsPoweredOn();
-    ui->edit_username->setText(QString::fromStdString(Settings::values.username));
+
     ui->combo_language->setCurrentIndex(Settings::values.language_index);
+
+    item_model->removeRows(0, item_model->rowCount());
+    list_items.clear();
+
+    std::transform(Settings::values.users.begin(), Settings::values.users.end(),
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: `<algorithm>` needs to be included.

**Comment by Subv** [comment]
Where is the list of users actually stored in the switch? I'm of the opinion that we should try to keep as close compatibility as possible with the real system so following that convention would be preferable.

**Comment by ogniK5377** [comment]
> Where is the list of users actually stored in the switch? I'm of the opinion that we should try to keep as close compatibility as possible with the real system so following that convention would be preferable.

The users are stored in a system save, as of currently we have no way of openning the save with our current backend and requires a lot of work. However in the future we should switch from the config -> the official syssave at some point

**Comment by DarkLordZach** [comment]
@Subv fixed

File: src/core/hle/service/am/am.cpp [hunk 21]
```diff
@@ -680,8 +682,10 @@ void IApplicationFunctions::PopLaunchParameter(Kernel::HLERequestContext& ctx) {
     std::vector<u8> buffer(POP_LAUNCH_PARAMETER_BUFFER_SIZE);
 
     std::memcpy(buffer.data(), header_data.data(), header_data.size());
-    const auto current_uuid = Settings::values.users[Settings::values.current_user].second.uuid;
-    std::memcpy(buffer.data() + header_data.size(), current_uuid.data(), sizeof(u128));
+
+    Account::ProfileManager profile_manager{};
+    const auto uuid = profile_manager.GetAllUsers()[Settings::values.current_user].uuid;
```
Comment (Subv) [inline src/core/hle/service/am/am.cpp]: Again, don't we have a GetUser? I'll stop pointing this out.

File: src/yuzu/configuration/configure_system.cpp [hunk 22]
```diff
@@ -99,49 +112,69 @@ void ConfigureSystem::setConfiguration() {
...
-                       }
-
-                       return QList{new QStandardItem{
-                           icon.scaled(64, 64, Qt::IgnoreAspectRatio, Qt::SmoothTransformation),
-                           QString::fromStdString(user.first + "\n" + user.second.Format())}};
-                   });
-
-    for (const auto& item : list_items)
-        item_model->appendRow(item);
+    ui->pm_add->setEnabled(profile_manager->GetUserCount() < 8);
```
Comment (Subv) [inline src/yuzu/configuration/configure_system.cpp]: what is 8?

File: src/yuzu/configuration/config.cpp [hunk 23]
```diff
@@ -123,23 +124,7 @@ void Config::ReadValues() {
...
-    }
-
-    qt_config->endArray();
-
-    if (Settings::values.users.empty())
-        Settings::values.users.emplace_back("yuzu", Service::Account::UUID{}.Generate());
-
-    Settings::values.current_user =
-        std::clamp(qt_config->value("current_user", 0).toInt(), 0, size);
+    Settings::values.current_user = std::clamp(qt_config->value("current_user", 0).toInt(), 0, 7);
```
Comment (Subv) [inline src/yuzu/configuration/config.cpp]: don't use magic numbers, `ProfileManager::MaxUsers - 1`

File: src/yuzu/configuration/configure_system.cpp [hunk 24]
```diff
@@ -99,49 +112,69 @@ void ConfigureSystem::setConfiguration() {
...
+            const auto username = Common::StringFromFixedZeroTerminatedBuffer(
+                reinterpret_cast<const char*>(profile.username.data()), profile.username.size());
+
+            return QList<QStandardItem*>{new QStandardItem{
+                GetIcon(user).scaled(64, 64, Qt::IgnoreAspectRatio, Qt::SmoothTransformation),
+                QString::fromStdString(username + '\n' + user.FormatSwitch())}};
+        });
+
+    list_items.erase(
+        std::remove_if(list_items.begin(), list_items.end(),
```
Comment (Subv) [inline src/yuzu/configuration/configure_system.cpp]: what is this doing and why? Add a comment

File: src/yuzu/configuration/configure_system.cpp [hunk 25]
```diff
@@ -99,49 +112,69 @@ void ConfigureSystem::setConfiguration() {
...
+        icon.fill(Qt::black);
         icon.loadFromData(backup_jpeg.data(), backup_jpeg.size());
     }
 
+    return icon;
+}
+
+void ConfigureSystem::PopulateUserList() {
+    const auto& profiles = profile_manager->GetAllUsers();
+    std::transform(
```
Comment (Subv) [inline src/yuzu/configuration/configure_system.cpp]: You're not doing a transform operation, you're doing a filter+transform, IMO it's better to have this as a range-based for

File: src/yuzu/configuration/configure_system.cpp [hunk 26]
```diff
@@ -212,44 +245,96 @@ void ConfigureSystem::AddUser() {
...
 
-    setConfiguration();
+    item_model->appendRow(new QStandardItem{
+        GetIcon(uuid).scaled(64, 64, Qt::IgnoreAspectRatio, Qt::SmoothTransformation),
+        QString::fromStdString(username.toStdString() + '\n' + uuid.FormatSwitch())});
 }
 
 void ConfigureSystem::RenameUser() {
     const auto user = tree_view->currentIndex().row();
+    ASSERT(user < 8);
```
Comment (Subv) [inline src/yuzu/configuration/configure_system.cpp]: 8?

File: src/yuzu/configuration/configure_system.cpp [hunk 27]
```diff
@@ -212,44 +245,96 @@ void ConfigureSystem::AddUser() {
...
+        QInputDialog::getText(this, tr("Enter Username"), tr("Enter a new username:"),
+                              QLineEdit::Normal, QString::fromStdString(username), &ok);
 
     if (!ok)
         return;
 
-    Settings::values.users[user].first = new_username.toStdString();
+    const auto username_std = new_username.toStdString();
+    if (username_std.size() > profile.username.size())
+        std::copy_n(username_std.begin(), profile.username.size(), profile.username.begin());
```
Comment (Subv) [inline src/yuzu/configuration/configure_system.cpp]: `std::copy_n(username_std.begin(), std::min(profile.username.size(), username_std.size()), profile.username.begin());`

File: src/yuzu/main.cpp [hunk 28]
```diff
@@ -726,10 +727,22 @@ void GMainWindow::OnGameListOpenFolder(u64 program_id, GameListOpenTarget target
...
         ASSERT(program_id != 0);
 
-        QStringList list{};
-        std::transform(Settings::values.users.begin(), Settings::values.users.end(),
-                       std::back_inserter(list),
-                       [](const auto& user) { return QString::fromStdString(user.first); });
+        Service::Account::ProfileManager manager{};
+        const auto user_ids = manager.GetAllUsers();
+        QStringList list;
+        std::transform(
```
Comment (Subv) [inline src/yuzu/main.cpp]: Again, you're not performing a map but a filter+map, this would be better off as a range-based loop

File: src/core/hle/service/acc/profile_manager.cpp [hunk 29]
```diff
@@ -241,4 +264,70 @@ bool ProfileManager::CanSystemRegisterUser() const {
...
 }
 
+bool ProfileManager::RemoveUser(UUID uuid) {
+    auto index = GetUserIndex(uuid);
+    if (index == boost::none) {
+        return false;
+    }
+
+    profiles[*index] = ProfileInfo{};
+    std::stable_partition(profiles.begin(), profiles.end(),
```
[the remark this answers is not in the document]
Comment (DarkLordZach) [reply src/core/hle/service/acc/profile_manager.cpp]: If the user was not the last user, it has to be put at the back. From my understanding, both the class and switch operate on the property that the users array has all of the users with invalid UUIDs (0) at the end.

File: src/yuzu/configuration/configure_system.h [hunk 30]
```diff
@@ -5,7 +5,16 @@
 #pragma once
 
 #include <memory>
+
+#include <QList>
 #include <QWidget>
+#include "core/hle/service/acc/profile_manager.h"
+
+class QVBoxLayout;
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.h]: These should be alphabetically ordered.

File: src/yuzu/configuration/configure_system.cpp [hunk 31]
```diff
@@ -24,31 +40,141 @@ static const std::array<int, 12> days_in_month = {{
...
+
+    scene->clear();
+    scene->addPixmap(
+        GetIcon(*current_user).scaled(48, 48, Qt::IgnoreAspectRatio, Qt::SmoothTransformation));
+    ui->current_user_username->setText(QString::fromStdString(username));
 }
 
 void ConfigureSystem::ReadSystemSettings() {}
 
+std::string ConfigureSystem::GetAccountUsername(Service::Account::UUID uuid) {
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: This can be a const member function

File: src/core/hle/service/acc/profile_manager.cpp [hunk 32]
```diff
@@ -239,4 +278,76 @@ bool ProfileManager::CanSystemRegisterUser() const {
...
+    return true;
+}
+
+void ProfileManager::ParseUserSaveFile() {
+    FileUtil::IOFile save(FileUtil::GetUserPath(FileUtil::UserPath::NANDDir) +
+                              ACC_SAVE_AVATORS_BASE_PATH + "profiles.dat",
+                          "rb");
+
+    ProfileDataRaw data;
+    save.Seek(0, SEEK_SET);
```
Comment (lioncash) [inline src/core/hle/service/acc/profile_manager.cpp]: You don't need to seek to the beginning on an immediately opened file. It's already at offset 0.

File: src/core/hle/service/acc/profile_manager.cpp [hunk 33]
```diff
@@ -239,4 +278,76 @@ bool ProfileManager::CanSystemRegisterUser() const {
...
+    auto& profile = profiles[*index];
+    profile.user_uuid = profile_new.user_uuid;
+    profile.username = profile_new.username;
+    profile.creation_time = profile_new.timestamp;
+
+    return true;
+}
+
+void ProfileManager::ParseUserSaveFile() {
+    FileUtil::IOFile save(FileUtil::GetUserPath(FileUtil::UserPath::NANDDir) +
```
Comment (lioncash) [inline src/core/hle/service/acc/profile_manager.cpp]: There should be handling code for this not opening properly.

File: src/core/hle/service/acc/profile_manager.cpp [hunk 34]
```diff
@@ -239,4 +278,76 @@ bool ProfileManager::CanSystemRegisterUser() const {
...
+    }
+
+    const auto raw_path =
+        FileUtil::GetUserPath(FileUtil::UserPath::NANDDir) + "/system/save/8000000000000010";
+    if (FileUtil::Exists(raw_path) && !FileUtil::IsDirectory(raw_path))
+        FileUtil::Delete(raw_path);
+
+    const auto path = FileUtil::GetUserPath(FileUtil::UserPath::NANDDir) +
+                      ACC_SAVE_AVATORS_BASE_PATH + "profiles.dat";
+    FileUtil::CreateFullPath(path);
```
Comment (lioncash) [inline src/core/hle/service/acc/profile_manager.cpp]: This is something that can fail.

File: src/core/hle/service/acc/profile_manager.cpp [hunk 35]
```diff
@@ -239,4 +278,76 @@ bool ProfileManager::CanSystemRegisterUser() const {
...
+    FileUtil::IOFile save(path, "wb");
```
Comment (lioncash) [inline src/core/hle/service/acc/profile_manager.cpp]: There should be code to handle this failing to open.

File: src/core/hle/service/acc/profile_manager.cpp [hunk 36]
```diff
@@ -239,4 +278,76 @@ bool ProfileManager::CanSystemRegisterUser() const {
...
+
+    save.Resize(sizeof(ProfileDataRaw));
+    save.Seek(0, SEEK_SET);
```
Comment (lioncash) [inline src/core/hle/service/acc/profile_manager.cpp]: This seek is unnecessary.

File: src/core/hle/service/acc/profile_manager.cpp [hunk 37]
```diff
@@ -23,13 +42,21 @@ const UUID& UUID::Generate() {
...
 
 ProfileManager::ProfileManager() {
-    // TODO(ogniK): Create the default user we have for now until loading/saving users is added
-    auto user_uuid = UUID{1, 0};
-    ASSERT(CreateNewUser(user_uuid, Settings::values.username).IsSuccess());
-    OpenUser(user_uuid);
+    ParseUserSaveFile();
+
+    if (user_count == 0)
+        CreateNewUser(UUID{}.Generate(), "yuzu");
```
Comment (lioncash) [inline src/core/hle/service/acc/profile_manager.cpp]: You may want to consider just making `Generate()` a static function and changing exiting usages to operate in terms of that, given this is pretty wonky.

File: src/core/hle/service/acc/profile_manager.cpp [hunk 38]
```diff
@@ -4,15 +4,34 @@
...
+    std::array<UserRaw, MAX_USERS> users;
+};
+static_assert(sizeof(ProfileDataRaw) == 0x650, "ProfileDataRaw has incorrect size.");
+
 // TODO(ogniK): Get actual error codes
 constexpr ResultCode ERROR_TOO_MANY_USERS(ErrorModule::Account, -1);
 constexpr ResultCode ERROR_USER_ALREADY_EXISTS(ErrorModule::Account, -2);
 constexpr ResultCode ERROR_ARGUMENT_IS_NULL(ErrorModule::Account, 20);
 
+constexpr const char* ACC_SAVE_AVATORS_BASE_PATH = "/system/save/8000000000000010/su/avators/";
```
Comment (lioncash) [inline src/core/hle/service/acc/profile_manager.cpp]:
```cpp
constexpr char ACC_SAVE_AVATORS_BASE_PATH[] = "/system/save/8000000000000010/su/avators/";
```

Is more space-efficient, given it only stores the string data. The previous code stores both the string data and a pointer address into the binary.

File: src/yuzu/configuration/configure_system.cpp [hunk 39]
```diff
@@ -92,3 +218,125 @@ void ConfigureSystem::refreshConsoleID() {
...
+    if (!ok)
+        return;
+
+    const auto username_std = new_username.toStdString();
+    if (username_std.size() > profile.username.size()) {
+        std::copy_n(username_std.begin(), std::min(profile.username.size(), username_std.size()),
+                    profile.username.begin());
+    } else {
+        std::copy(username_std.begin(), username_std.end(), profile.username.begin());
+    }
```
Comment (kyn3tic) [inline src/yuzu/configuration/configure_system.cpp]: I don't believe this properly handles when the new name is shorter than the existing name. After renaming with a short name, the resulting name retains its original length with only the first n characters replaced. The list UI will however show the new name correctly until it's reopened.

Comment (DarkLordZach) [reply src/yuzu/configuration/configure_system.cpp]: Fixed in 1c5885d by clearing the username array before copying the new name in.

File: src/yuzu/configuration/configure_system.cpp [hunk 40]
```diff
@@ -92,3 +218,126 @@ void ConfigureSystem::refreshConsoleID() {
...
+
+    profile_manager->SetProfileBase(*uuid, profile);
+
+    item_model->setItem(
+        user, 0,
+        new QStandardItem{
+            GetIcon(*uuid).scaled(64, 64, Qt::IgnoreAspectRatio, Qt::SmoothTransformation),
+            QString::fromStdString(username_std + '\n' + uuid->FormatSwitch())});
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: This should probably be a translatable string like:

```cpp
tr("%1\n%2", 
   "%1 is the profile username, %2 is the formatted UUID (e.g. 00112233-4455-6677-8899-AABBCCDDEEFF))")
        .arg(QString::fromStdString(username_std), QString::fromStdString(uuid->FormatSwitch()));
```

so translators have control over the text flow.

File: src/yuzu/configuration/configure_system.cpp [hunk 41]
```diff
@@ -92,3 +218,126 @@ void ConfigureSystem::refreshConsoleID() {
...
+
+void ConfigureSystem::DeleteUser() {
+    const auto index = tree_view->currentIndex().row();
+    const auto uuid = profile_manager->GetUser(index);
+    ASSERT(uuid != boost::none);
+    const auto username = GetAccountUsername(*uuid);
+
+    const auto confirm = QMessageBox::question(
+        this, tr("Confirm Delete"),
+        tr("You are about to delete user with name %1. Are you sure?").arg(username.c_str()));
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.cpp]: `.arg(QString::fromStdString(username))`

File: src/yuzu/configuration/configure_system.h [hunk 42]
```diff
@@ -5,7 +5,16 @@
 #pragma once
 
 #include <memory>
+
+#include <QList>
 #include <QWidget>
+#include "core/hle/service/acc/profile_manager.h"
```
Comment (lioncash) [inline src/yuzu/configuration/configure_system.h]: This can be replaced with forward declarations of

```cpp
namespace Service::Account {
class ProfileManager;
struct UUID;
};
```

and have the include be moved into the cpp file.
